.grit/patterns/python/cleaned_data_vs_post.md

```diff
@@ -0,0 +1,113 @@
+---
+title: Use `$FORM.cleaned_data[]` instead of `request.POST[]` after `form.is_valid()`
+tags: [fix, best-practice]
+---
+
+Use `$FORM.cleaned_data[]` instead of `request.POST[]` after `form.is_valid()` has been executed to only access sanitized data.
+
+### references
+
+- https://docs.djangoproject.com/en/4.2/ref/forms/api/#accessing-clean-data
+
+tags: #fix, #best-practice
+
+```grit
+engine marzano(0.1)
+language python
+
+`if $form.is_valid(): $conditionBody` where {
+    $conditionBody <: contains `$request.POST` => `$form.cleaned_data`
+}
+```
+
+## with `request.POST`
+
+```python
+from django.shortcuts import render, redirect
+from .models import *
+from .forms import *
+
+def create_new_tournament_dangerous(request):
+    if request.method == 'POST':
+        form = CreateTournamentForm(request.POST)
+        if form.is_valid():
+            t = Tournament(name=request.POST['name'])
+            t.save()
+            return redirect('index')
+    else:
+        context = { 'form': CreateTournamentForm()}
+        return render(request, 'create_tournament.html', context)
+```
+
+```python
+from django.shortcuts import render, redirect
+from .models import *
+from .forms import *
+
+def create_new_tournament_dangerous(request):
+    if request.method == 'POST':
+        form = CreateTournamentForm(request.POST)
+        if form.is_valid():
+            t = Tournament(name=form.cleaned_data['name'])
+            t.save()
+            return redirect('index')
+    else:
+        context = { 'form': CreateTournamentForm()}
+        return render(request, 'create_tournament.html', context)
+```
+
+## with `form.cleaned_data`
+
+```python
+def create_new_tournament_safe(request):
+    if request.method == 'POST':
+        form = CreateTournamentForm(request.POST)
+        if form.is_valid():
+            t = Tournament(name=form.cleaned_data['name'])
+            t.save()
+            return redirect('index')
+    else:
+        context = { 'form': CreateTournamentForm()}
+        return render(request, 'create_tournament.html', context)
+```
+
+```python
+def create_new_tournament_safe(request):
+    if request.method == 'POST':
+        form = CreateTournamentForm(request.POST)
+        if form.is_valid():
+            t = Tournament(name=form.cleaned_data['name'])
+            t.save()
+            return redirect('index')
+    else:
+        context = { 'form': CreateTournamentForm()}
+        return render(request, 'create_tournament.html', context)
+```
+
+## with `request.POST.get`
+
+```python
+def create_new_tournament_dangerous(request):
+    if request.method == 'POST':
+        form = CreateTournamentForm(request.POST)
+        if form.is_valid():          
+            t.save()
+            t = Tournament(name=request.POST.get['address'])
```

Comment:
Are you sure this is valid Django code? It looks off to me. Please make sure all samples you use are valid/real code.

Reply:
Hi @morgante, I think the importswas missing, updated it now

Reply:
Can you please check actual examples, or pull from a real Django example?

I'm pretty sure `.POST` is a dictionary. `.get[$index]` isn't valid on a dictionary.

This looks more correct:
```suggestion
            t = Tournament(name=request.POST.get('address'))
```

This gives me low confidence in the provided samples. Please ensure all test cases you provide come from real Django examples.

Reply:
Hi @morgante, my bad you where, correct checked examples, and found `t = Tournament(name=request.POST.get('address'))` this one is correct syntax, now this is good.